### DISCUSSION TITLE
Performance enhancements

### DISCUSSION THREAD

**Discussion by therampagerado** [thirtybees/ThirtyBees#1679] [Ideas]
Hello team,

Inspired by a recent find here is a short list of ideas for Performance controller enhancement. Could we implement some of them. What is the expected monetary cost of implementing such changes in the core? If viable we can set up crowdfund or gather a team of merchants that can fund those?

**Lazy loading:**
-lazy load images
-lazy load iframes
-'lazy load' videos (preload="none")

**Resources:**
-preload fonts (links to fonts here)
-preconnect to origins (links to resources here)
-prefetch on hover

**Page optimizations:**
-use passive event listeners
-http2 CSS push
-noopener tag to _blank links

**Code optimization:**
-defer javascript
-async image decoding
-add missing image size

**System improvements suggestions:**
-enable SSL on all pages
-check if smarty is active
-caching type - opcache recommended
-check if CCC items are YES
-check if debug mode is active
-check if profiling is active
-check if maintenance is active (link to controller so quick change is possible) or direct switch here
-check if server side cache is active (recommendations for apc, memcached, memcache or redis)

---alert in BO header if any of those are false and link to the Performance page->System Improvement suggestions part so merchant can fix quickly

**Server configuration:**
-display server configuration parameters together with recommended values. Information that merchant can tweak some of them or he/she must contact host to fix for them (if possible)
-PHP version - same information
-PHP extension list - list of needed extensions by core/availability/information

**Database Configuration:**
-display DB settings and recommended values - shows information how to configure those if own server or contact host to tweak them (if possible)

**Comment by getdatakick** [comment]
Most of this is a  responsibility of theme(s). We could, of course, implement them in default themes, but that will not help people using third party themes.

What we could do in core

- http2 CSS push
- preconnect to origins. Maybe implement as a separate module
- preload fonts. We would have to parse the css files to resolve dependencies. Again, maybe module would be better than adding this to core
- Alert in BO for System improvements suggestions

**Comment by therampagerado** [comment]
Yes, you're right most of those are heavily theme related but could we override them on system level? The 2 most popular 3rd party themes compatible with TB are frozen in time and if we can improve on those?

Few more low-hanging items:
- increasing expire time on static resources (all speed tests complain for longer periods). I tested with 1 month for css/js and 1 year for images and the tools appear to be OK with those? Your opinion what is optimal here?

I believe all speed improvements that can be put in the core should be in one place. Yes, sometimes not everything will be needed (conflicts, etc) but if the merchant sees the option it will make them use it more readily. If we have them as modules for this, module for that, yes it will be good for fragmentation not bulking up the core but it will leave this code unused outside the core of the TB community that follows the development closely (as with the email transport modules where you had to put notices).

**Comment by therampagerado** [comment]
And of course probably the biggest improvement will be the [rework of the images section ](https://github.com/thirtybees/thirtybees/tree/clean-webp-support)- currently (for me) this is the largest delay and it will bring the biggest speed improvement.
